### PR TITLE
rgw: improve buffer list utilization in the chunkupload scenario

### PR DESCRIPTION
**Pull request by VVoidV**
Reduced waste of buffer::ptr by receiving multiple chunks and filling them into the buffer

AWSv4ComplMulti::recv_body() just receive one chunk and fill it into buffer. Each 4MB(rgw_max_chunk_size)  buffer is actually only utilizing 64KB, leading to frequent buffer allocations. ~800GB virtual memory consumption has been observed.

https://tracker.ceph.com/issues/62681

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by VVoidV** [comment]
@rzarzynski   could you please take a look? 
top shows the virtual mem usage is super high 
image

image

File: src/rgw/rgw_auth_s3.cc [hunk 1]
```diff
@@ -1225,6 +1227,11 @@ size_t AWSv4ComplMulti::recv_body(char* const buf, const size_t buf_max)
     to_extract -= received;
   }
 
+  if(buf_pos < buf_max && received > 0) {
+    dout(30) << "AWSv4ComplMulti: get next chunk, total buffer:" << buf_max <<", filled=" << buf_pos << dendl;
+    goto again;
```
Comment (cbodley) [inline src/rgw/rgw_auth_s3.cc:1232]: thanks @VVoidV, i definitely agree that we should loop until we've filled the input buffer. the use of `goto` here does simplify the diff, but i'd prefer to use a real loop construct instead

what do you think about renaming this `AWSv4ComplMulti::recv_body()` function to `recv_chunk()` and adding a new `recv_body()` that loops over calls to `recv_chunk()`? i think the only necessary addition to `recv_chunk()` is a `bool& eof` argument that gets set in the `received == 0` cases

Comment (VVoidV) [reply src/rgw/rgw_auth_s3.cc]: Thanks @cbodley, the loop version is done.  Please take a look.

**Review by cbodley — COMMENTED** [review]
@VVoidV looks great! can you confirm that the updated version still improves the memory utilization?

**Comment by VVoidV** [comment]
> @VVoidV looks great! can you confirm that the updated version still improves the memory utilization?

Yes,  run the warp test on my vstart cluster with command below.

```
warp put --host=127.0.0.1:8000 --bucket=warp-multi-write-100m --access-key=0555b35654ad1656d804 --secret-key=h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q==  --obj.randsize --obj.size=200Mib --duration=1m --concurrent=16 --benchdata=200M-multi-write -q --disable-multipart --noclear
```

With this patch, VIRT is abort **5GB**

image

without this patch, VIRT is abort **18GB**

image



It also passed the test with --md5 flag. RGW will compare suppiled_md5 with the MD5 calculated from the received data.
```
# warp put --host=127.0.0.1:8000 --bucket=warp-multi-write-200m --access-key=0555b35654ad1656d804 --secret-key=h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q==  --obj.randsize --obj.size=200Mib --duration=1m --concurrent=64 --benchdata=200M-multi-write -q --disable-multipart --md5
warp: Benchmark data written to "200M-multi-write.csv.zst"                                                                                                  

----------------------------------------
Operation: PUT
* Average: 456.47 MiB/s, 12.94 obj/s

Throughput, split into 5 x 1s:
 * Fastest: 464.8MiB/s, 9.26 obj/s
 * 50% Median: 460.0MiB/s, 12.33 obj/s
 * Slowest: 444.8MiB/s, 17.43 obj/s
```

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2023-09-29_14:54:02-rgw-wip-cbodley-testing-distro-default-smithi/

**Comment by microyahoo** [comment]
hi @VVoidV @cbodley, this PR will cause an error when uploading empty file using the aws Java SDK. For specific reproduction steps, see the [s3-java-demo](https://github.com/microyahoo/s3-java-demo). Seem like the while loop has been skipped.
https://github.com/ceph/ceph/blame/618f440892089921c3e944a991122ddc44e60516/src/rgw/rgw_auth_s3.cc#L1114

```console
2024-09-02T06:26:58.150+0000 7fd81c389700  1 ====== req done req=0x7fd8aacf5600 op status=0 http_status=200 latency=0.001000008s ======
2024-09-02T06:26:58.150+0000 7fd81c389700  1 beast: 0x7fd8aacf5600: 10.9.8.72 - - [02/Sep/2024:06:26:58.149 +0000] "GET /swift/healthcheck HTTP/1.1" 200 0 - "kube-probe/1.22" - latency=0.001000008s
2024-09-02T06:27:01.250+0000 7fd87d44b700  2 rgw data changes log: RGWDataChangesLog::ChangesRenewThread: start
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 CONTENT_LENGTH=86
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 CONTENT_TYPE=application/octet-stream
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_AMZ_SDK_INVOCATION_ID=ffc563d8-0fb2-9653-9cff-ded422c3c4f3
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_AMZ_SDK_REQUEST=attempt=1; max=4
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_AUTHORIZATION=AWS4-HMAC-SHA256 Credential=testy/20240902/us-east-1/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;content-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-decoded-content-length, Signature=9ef62b67fde6b9f8155cfffb008384fee62bb5df9e5056e591b7b888dd450b46
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_CONNECTION=Keep-Alive
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_EXPECT=100-continue
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_HOST=10.9.8.72
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_USER_AGENT=aws-sdk-java/2.20.43 Linux/3.10.0-1127.el7.x86_64 OpenJDK_64-Bit_Server_VM/25.412-b08 Java/1.8.0_412 vendor/Red_Hat__Inc. io/sync http/Apache cfg/retry-mode/legacy
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_VERSION=1.1
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_X_AMZ_CONTENT_SHA256=STREAMING-AWS4-HMAC-SHA256-PAYLOAD
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_X_AMZ_DATE=20240902T062703Z
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 HTTP_X_AMZ_DECODED_CONTENT_LENGTH=0
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 REMOTE_ADDR=10.9.8.72
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 REQUEST_METHOD=PUT
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 REQUEST_URI=/test/chongyin-test-001
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 SCRIPT_URI=/test/chongyin-test-001
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 SERVER_PORT=80
2024-09-02T06:27:03.634+0000 7fd81bb88700  1 ====== starting new request req=0x7fd8aacf5600 =====
2024-09-02T06:27:03.634+0000 7fd81bb88700  2 req 17626438537029431592 0.000000000s initializing for trans_id = tx00000f49db075e6e53128-0066d55ab7-107aa6d-object-store
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s rgw api priority: s3=8 s3website=7
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s host=10.9.8.72
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 req 17626438537029431592 0.000000000s subdomain= domain= in_hosted_domain=0 in_hosted_domain_s3website=0
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 req 17626438537029431592 0.000000000s final domain/bucket subdomain= domain= in_hosted_domain=0 in_hosted_domain_s3website=0 s->info.domain= s->info.request_uri=/test/chongyin-test-001
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s meta>> HTTP_X_AMZ_CONTENT_SHA256
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s meta>> HTTP_X_AMZ_DATE
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s meta>> HTTP_X_AMZ_DECODED_CONTENT_LENGTH
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s x>> x-amz-content-sha256:STREAMING-AWS4-HMAC-SHA256-PAYLOAD
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s x>> x-amz-date:20240902T062703Z
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s x>> x-amz-decoded-content-length:0
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 req 17626438537029431592 0.000000000s get_handler handler=22RGWHandler_REST_Obj_S3
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s handler=22RGWHandler_REST_Obj_S3
2024-09-02T06:27:03.634+0000 7fd81bb88700  2 req 17626438537029431592 0.000000000s getting op 1
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s s3:put_obj scheduling with throttler client=2 cost=1
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s s3:put_obj op=21RGWPutObj_ObjStore_S3
2024-09-02T06:27:03.634+0000 7fd81bb88700  2 req 17626438537029431592 0.000000000s s3:put_obj verifying requester
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 req 17626438537029431592 0.000000000s s3:put_obj rgw::auth::StrategyRegistry::s3_main_strategy_t: trying rgw::auth::s3::AWSAuthStrategy
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 req 17626438537029431592 0.000000000s s3:put_obj rgw::auth::s3::AWSAuthStrategy: trying rgw::auth::s3::S3AnonymousEngine
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 req 17626438537029431592 0.000000000s s3:put_obj rgw::auth::s3::S3AnonymousEngine denied with reason=-1
2024-09-02T06:27:03.634+0000 7fd81bb88700 20 req 17626438537029431592 0.000000000s s3:put_obj rgw::auth::s3::AWSAuthStrategy: trying rgw::auth::s3::LocalEngine
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s v4 signature format = 9ef62b67fde6b9f8155cfffb008384fee62bb5df9e5056e591b7b888dd450b46
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s v4 credential format = testy/20240902/us-east-1/s3/aws4_request
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s access key id = testy
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s credential scope = 20240902/us-east-1/s3/aws4_request
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s canonical headers format = amz-sdk-invocation-id:ffc563d8-0fb2-9653-9cff-ded422c3c4f3
amz-sdk-request:attempt=1; max=4
content-length:86
content-type:application/octet-stream
host:10.9.8.72
x-amz-content-sha256:STREAMING-AWS4-HMAC-SHA256-PAYLOAD
x-amz-date:20240902T062703Z
x-amz-decoded-content-length:0

2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s payload request hash = STREAMING-AWS4-HMAC-SHA256-PAYLOAD
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s canonical request = PUT
/test/chongyin-test-001

amz-sdk-invocation-id:ffc563d8-0fb2-9653-9cff-ded422c3c4f3
amz-sdk-request:attempt=1; max=4
content-length:86
content-type:application/octet-stream
host:10.9.8.72
x-amz-content-sha256:STREAMING-AWS4-HMAC-SHA256-PAYLOAD
x-amz-date:20240902T062703Z
x-amz-decoded-content-length:0

amz-sdk-invocation-id;amz-sdk-request;content-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-decoded-content-length
STREAMING-AWS4-HMAC-SHA256-PAYLOAD
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s canonical request hash = 3d8e9f103c95e4c32d80183f373dacead6637ab182dae81bc33beb330aa405f0
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s string to sign = AWS4-HMAC-SHA256
20240902T062703Z
20240902/us-east-1/s3/aws4_request
3d8e9f103c95e4c32d80183f373dacead6637ab182dae81bc33beb330aa405f0
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s body content detected in multiple chunks
2024-09-02T06:27:03.634+0000 7fd81bb88700 10 req 17626438537029431592 0.000000000s aws4 seed signature ok... delaying v4 auth
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s date_k    = 6530246fcca25cd46da2128f91f5163a460cbc73096a977be2d3e8db3ce6e8ae
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s region_k  = 6778fadfde7930258fc26c9b9ae63849a42647b01e884b6f91abe243436679a0
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s service_k = 9748c20816e8704913bba2332dbd281aad4c53e3e640b3ba79a4e33cf7527db5
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s signing_k = 6302a37eb916b95f60071c36c14469ec5bee490ce4e70fba14d3e3e3e71babbb
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s generated signature = 9ef62b67fde6b9f8155cfffb008384fee62bb5df9e5056e591b7b888dd450b46
2024-09-02T06:27:03.635+0000 7fd81bb88700 15 req 17626438537029431592 0.001000008s s3:put_obj string_to_sign=AWS4-HMAC-SHA256
20240902T062703Z
20240902/us-east-1/s3/aws4_request
3d8e9f103c95e4c32d80183f373dacead6637ab182dae81bc33beb330aa405f0
2024-09-02T06:27:03.635+0000 7fd81bb88700 15 req 17626438537029431592 0.001000008s s3:put_obj server signature=9ef62b67fde6b9f8155cfffb008384fee62bb5df9e5056e591b7b888dd450b46
2024-09-02T06:27:03.635+0000 7fd81bb88700 15 req 17626438537029431592 0.001000008s s3:put_obj client signature=9ef62b67fde6b9f8155cfffb008384fee62bb5df9e5056e591b7b888dd450b46
2024-09-02T06:27:03.635+0000 7fd81bb88700 15 req 17626438537029431592 0.001000008s s3:put_obj compare=0
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s date_k    = 6530246fcca25cd46da2128f91f5163a460cbc73096a977be2d3e8db3ce6e8ae
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s region_k  = 6778fadfde7930258fc26c9b9ae63849a42647b01e884b6f91abe243436679a0
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s service_k = 9748c20816e8704913bba2332dbd281aad4c53e3e640b3ba79a4e33cf7527db5
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s signing_k = 6302a37eb916b95f60071c36c14469ec5bee490ce4e70fba14d3e3e3e71babbb
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj rgw::auth::s3::LocalEngine granted access
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj rgw::auth::s3::AWSAuthStrategy granted access
2024-09-02T06:27:03.635+0000 7fd81bb88700  2 req 17626438537029431592 0.001000008s s3:put_obj normalizing buckets and tenants
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s s->object=chongyin-test-001 s->bucket=test
2024-09-02T06:27:03.635+0000 7fd81bb88700  2 req 17626438537029431592 0.001000008s s3:put_obj init permissions
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj get_system_obj_state: rctx=0x7fd8aacf3f80 obj=object-store.rgw.meta:root:test state=0x558451d852a0 s->prefetch_data=0
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj cache get: touching lru, lru_counter=201100 promotion_ts=200980
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s s3:put_obj moving object-store.rgw.meta+root+test to cache LRU end
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s s3:put_obj cache get: name=object-store.rgw.meta+root+test : hit (requested=0x16, cached=0x17)
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj get_system_obj_state: s->obj_tag was set empty
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s s3:put_obj cache get: name=object-store.rgw.meta+root+test : hit (requested=0x11, cached=0x17)
2024-09-02T06:27:03.635+0000 7fd81bb88700 15 req 17626438537029431592 0.001000008s s3:put_obj decode_policy Read AccessControlPolicy<AccessControlPolicy xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>test1</ID><DisplayName>test1</DisplayName></Owner><AccessControlList><Grant><Grantee xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="CanonicalUser"><ID>test1</ID><DisplayName>test1</DisplayName></Grantee><Permission>FULL_CONTROL</Permission></Grant></AccessControlList></AccessControlPolicy>
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj get_system_obj_state: rctx=0x7fd8aacf3a18 obj=object-store.rgw.meta:users.uid:test1 state=0x558451d852a0 s->prefetch_data=0
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj cache get: touching lru, lru_counter=201101 promotion_ts=200984
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s s3:put_obj moving object-store.rgw.meta+users.uid+test1 to cache LRU end
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s s3:put_obj cache get: name=object-store.rgw.meta+users.uid+test1 : hit (requested=0x6, cached=0x17)
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj get_system_obj_state: s->obj_tag was set empty
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj Read xattr: user.rgw.idtag
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s s3:put_obj cache get: name=object-store.rgw.meta+users.uid+test1 : hit (requested=0x3, cached=0x17)
2024-09-02T06:27:03.635+0000 7fd81bb88700  2 req 17626438537029431592 0.001000008s s3:put_obj recalculating target
2024-09-02T06:27:03.635+0000 7fd81bb88700  2 req 17626438537029431592 0.001000008s s3:put_obj reading permissions
2024-09-02T06:27:03.635+0000 7fd81bb88700  2 req 17626438537029431592 0.001000008s s3:put_obj init op
2024-09-02T06:27:03.635+0000 7fd81bb88700  2 req 17626438537029431592 0.001000008s s3:put_obj verifying op mask
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj required_mask= 2 user.op_mask=7
2024-09-02T06:27:03.635+0000 7fd81bb88700  2 req 17626438537029431592 0.001000008s s3:put_obj verifying op permissions
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 req 17626438537029431592 0.001000008s s3:put_obj -- Getting permissions begin with perm_mask=50
2024-09-02T06:27:03.635+0000 7fd81bb88700  5 req 17626438537029431592 0.001000008s s3:put_obj Searching permissions for identity=rgw::auth::SysReqApplier -> rgw::auth::LocalApplier(acct_user=test1, acct_name=test1, subuser=, perm_mask=15, is_admin=0) mask=50
2024-09-02T06:27:03.635+0000 7fd81bb88700  5 req 17626438537029431592 0.001000008s s3:put_obj Searching permissions for uid=test1
2024-09-02T06:27:03.635+0000 7fd81bb88700  5 req 17626438537029431592 0.001000008s s3:put_obj Found permission: 15
2024-09-02T06:27:03.635+0000 7fd81bb88700  5 req 17626438537029431592 0.001000008s s3:put_obj Searching permissions for group=1 mask=50
2024-09-02T06:27:03.635+0000 7fd81bb88700  5 req 17626438537029431592 0.001000008s s3:put_obj Permissions for group not found
2024-09-02T06:27:03.635+0000 7fd81bb88700  5 req 17626438537029431592 0.001000008s s3:put_obj Searching permissions for group=2 mask=50
2024-09-02T06:27:03.635+0000 7fd81bb88700  5 req 17626438537029431592 0.001000008s s3:put_obj Permissions for group not found
2024-09-02T06:27:03.635+0000 7fd81bb88700  5 req 17626438537029431592 0.001000008s s3:put_obj -- Getting permissions done for identity=rgw::auth::SysReqApplier -> rgw::auth::LocalApplier(acct_user=test1, acct_name=test1, subuser=, perm_mask=15, is_admin=0), owner=test1, perm=2
2024-09-02T06:27:03.635+0000 7fd81bb88700 10 req 17626438537029431592 0.001000008s s3:put_obj  identity=rgw::auth::SysReqApplier -> rgw::auth::LocalApplier(acct_user=test1, acct_name=test1, subuser=, perm_mask=15, is_admin=0) requested perm (type)=2, policy perm=2, user_perm_mask=2, acl perm=2
2024-09-02T06:27:03.635+0000 7fd81bb88700  2 req 17626438537029431592 0.001000008s s3:put_obj verifying op params
2024-09-02T06:27:03.635+0000 7fd81bb88700  2 req 17626438537029431592 0.001000008s s3:put_obj pre-executing
2024-09-02T06:27:03.635+0000 7fd81bb88700  2 req 17626438537029431592 0.001000008s s3:put_obj executing
2024-09-02T06:27:03.635+0000 7fd81bb88700 20 rgw main: initiating async quota refresh for user=test1
2024-09-02T06:27:03.639+0000 7fd89647d700 20 async stats refresh response for bucket=:test[dddfd3cc-7378-448d-9c59-2983d33bce83.46097.2])
2024-09-02T06:27:03.640+0000 7fd80a365700 20 user quota: max_objects=-1 max_size=-1
2024-09-02T06:27:03.640+0000 7fd80a365700 20 user quota OK: stats.num_objects=51785 stats.size=1394335482
2024-09-02T06:27:03.654+0000 7fd80a365700 20 req 17626438537029431592 0.020000160s s3:put_obj max_chunk_size=4194304
2024-09-02T06:27:03.654+0000 7fd80a365700 20 AWSv4ComplMulti: received=0
2024-09-02T06:27:03.654+0000 7fd80a365700  5 req 17626438537029431592 0.020000160s s3:put_obj NOTICE: call to do_aws4_auth_completion
2024-09-02T06:27:03.654+0000 7fd80a365700 20 AWSv4ComplMulti: string_to_sign=
AWS4-HMAC-SHA256-PAYLOAD
20240902T062703Z
20240902/us-east-1/s3/aws4_request
9ef62b67fde6b9f8155cfffb008384fee62bb5df9e5056e591b7b888dd450b46
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2024-09-02T06:27:03.654+0000 7fd80a365700 20 AWSv4ComplMulti: ERROR: chunk signature mismatch
2024-09-02T06:27:03.654+0000 7fd80a365700 20 AWSv4ComplMulti: declared signature=9ef62b67fde6b9f8155cfffb008384fee62bb5df9e5056e591b7b888dd450b46
2024-09-02T06:27:03.654+0000 7fd80a365700 20 AWSv4ComplMulti: calculated signature=bf84227ae8089c1bad452e49fd0c24627e7d5af867e2b8f5fb1c47aa986a1746
2024-09-02T06:27:03.654+0000 7fd80a365700 10 ERROR: signature of last chunk does not match
2024-09-02T06:27:03.654+0000 7fd80a365700 20 req 17626438537029431592 0.020000160s s3:put_obj get_data() returned ret=-2040
2024-09-02T06:27:03.654+0000 7fd80a365700  2 req 17626438537029431592 0.020000160s s3:put_obj completing
2024-09-02T06:27:03.655+0000 7fd80a365700  2 req 17626438537029431592 0.021000167s s3:put_obj op status=-2040
2024-09-02T06:27:03.655+0000 7fd80a365700  2 req 17626438537029431592 0.021000167s s3:put_obj http status=400
2024-09-02T06:27:03.655+0000 7fd80a365700  1 ====== req done req=0x7fd8aacf5600 op status=-2040 http_status=400 latency=0.021000167s ======
2024-09-02T06:27:03.655+0000 7fd80a365700  1 beast: 0x7fd8aacf5600: 10.9.8.72 - test1 [02/Sep/2024:06:27:03.634 +0000] "PUT /test/chongyin-test-001 HTTP/1.1" 400 267 - "aws-sdk-java/2.20.43 Linux/3.10.0-1127.el7.x86_64 OpenJDK_64-Bit_Server_VM/25.412-b08 Java/1.8.0_412 vendor/Red_Hat__Inc. io/sync http/Apache cfg/retry-mode/legacy" - latency=0.021000167s
```

ceph pacific v16.2.15 will fail
[ceph-v16.2.15.log](https://github.com/user-attachments/files/16835359/ceph-v16.2.15.log)


ceph pacific v16.2.14 is normal.
[ceph-v16.2.14.log](https://github.com/user-attachments/files/16835360/ceph-v16.2.14.log)

**Comment by microyahoo** [comment]
https://tracker.ceph.com/issues/67875

**Comment by microyahoo** [comment]
seems this issue has been fixed by https://github.com/ceph/ceph/pull/54856

**Pull request merged**